consensus/types.go

```diff
@@ -1,53 +1,221 @@
 package consensus
 
 import (
+	"sync"
 	"time"
 
 	"github.com/deso-protocol/core/bls"
+	"github.com/deso-protocol/core/collections/bitset"
+	"github.com/holiman/uint256"
 )
 
-// ConsensusEvent is a way for FastHotStuffConsensus to send messages back to
-// the Server. There are three types of events that can be sent:
-//   - Vote: The consensus is ready to vote on a block
-//   - Timeout: The consensus has timed out on a view
-//   - BlockProposal: The consensus has a QC for a block and is ready to propose the next
-//     block
+// FastHotStuffEventType is a way for FastHotStuffEventLoop to send messages back to the Server.
+// There are four types of events that can be sent:
+//   - FastHotStuffEventTypeVote: The event loop is ready to vote on a block at a given block height and view
+//   - FastHotStuffEventTypeTimeout: The event loop has timed out on a view
+//   - FastHotStuffEventTypeConstructVoteQC: The event loop has a QC for a block and is ready to construct the
+//     next block at the next block height and the current view
+//   - FastHotStuffEventTypeConstructTimeoutQC: The event loop has a timeout QC for a view and is ready to
+//     construct an empty block with the timeout QC at the next block height and the current view
 
-type ConsensusEventType byte
+type FastHotStuffEventType byte
 
 const (
-	ConsensusEventTypeTimeout ConsensusEventType = iota
-	ConsensusEventTypeBlockProposal
-	ConsensusEventTypeVote
+	FastHotStuffEventTypeVote               FastHotStuffEventType = 0
+	FastHotStuffEventTypeTimeout            FastHotStuffEventType = 1
+	FastHotStuffEventTypeConstructVoteQC    FastHotStuffEventType = 2
+	FastHotStuffEventTypeConstructTimeoutQC FastHotStuffEventType = 3
 )
 
-type ConsensusEvent struct {
-	EventType   ConsensusEventType
-	BlockHeight uint64
-	View        uint64
+type FastHotStuffEvent struct {
+	EventType      FastHotStuffEventType
+	TipBlockHash   BlockHash
+	TipBlockHeight uint64
+	View           uint64
+	QC             QuorumCertificate
+	AggregateQC    AggregateQuorumCertificate
+}
+
+// Create an alias type of the 32 bit block hash so that the raw [32]byte type isn't
+// ambiguously repeated in the code base
+type BlockHashValue = [32]byte
```

Comment:
We already have BlockHash aliased to this. Any reason you didn't want to use that here?

Reply:
Nevermind I think based on the comment below you want to decouple this type from lib. I just realized that none of this is in lib now.

Reply:
Correct. We want to decouple the this consensus package from lib. But we define the BlockHash as an interface in the consensus package so that the two can interop